server/service/service.go

```diff
@@ -269,29 +288,165 @@ func (s *Service) ReportStatus(ctx context.Context, req *bpb.ReportStatusRequest
 	return &bpb.EmptyResponse{}, s.em.SetStatus(ctx, req)
 }
 
+const (
+	stateInitial = iota
+	// TPM 1.2 states
+	stateTPM12ChallengeSent
+	stateTPM12EKIdentitySent
+	// TPM 2.0 states
+	stateTPM20CSRRequested
+	stateTPM20NonceSent
+	// Common state
+	stateAttested
+)
+
+// Function to check if a client cert was presented
+func hasClientCert(ctx context.Context) bool {
+	p, ok := peer.FromContext(ctx)
+	if !ok {
+		return false
+	}
+	tlsInfo, ok := p.AuthInfo.(credentials.TLSInfo)
+	if !ok {
+		return false
+	}
+	return len(tlsInfo.State.PeerCertificates) > 0
+}
+
 func (s *Service) BootstrapStream(stream bpb.Bootstrap_BootstrapStreamServer) error {
+	ctx := stream.Context()
+
+	if hasClientCert(ctx) {
+		return status.Errorf(codes.FailedPrecondition, "BootstrapStream is not for clients presenting client certificates in TLS; use GetBootstrapData/ReportStatus RPCs instead")
+	}
+
+	var deviceID string
+	session := &streamSession{currentState: stateInitial}
+	var err error
+
+	// --- Initial Message ---
+	in, err := stream.Recv()
```

Comment:
Would it be possible to remove this single call to stream.Recv() and instead just have everything in a constant loop (e.g. on line 427().